Specialized double-SHA256 with 64 byte inputs with SSE4.1 and AVX2

sipa opened the pull request:
This introduces a framework for specialized double-SHA256 with 64 byte inputs. 4 different implementations are provided:
* Generic C++ (reusing the normal SHA256 code)
* Specialized C++ for 64-byte inputs, but no special instructions
* 4-way using SSE4.1 intrinsics
* 8-way using AVX2 intrinsics

On my own system (AVX2 capable), I get these benchmarks for computing the Merkle root of 9001 leaves (supported lengths / special instructions / parallellism):
* 7.2 ms with varsize/naive/1way (master, non-SSE4 hardware)
* 5.8 ms with size64/naive/1way (this PR, non-SSE4 capable systems)
* 4.8 ms with varsize/SSE4/1way (master, SSE4 hardware)
* 2.9 ms with size64/SSE4/4way (this PR, SSE4 hardware)
* 1.1 ms with size64/AVX2/8way (this PR, AVX2 hardware)

sipa committed:
Specialized double sha256 for 64 byte inputs

sipa committed:
8-way AVX2 implementation for double SHA256 on 64-byte inputs

sipa committed:
Refactor SHA256 code

sipa committed:
Benchmark Merkle root computation

sipa committed:
Use SHA256D64 in Merkle root computation

sipa committed:
4-way SSE4.1 implementation for double SHA256 on 64-byte inputs

laanwj commented:
Looks like these whole chain of functions are unused after this, except in the merkle tests:
```
BlockMerkleBranch → ComputeMerkleBranch → MerkleComputation
ComputeMerkleRootFromBranch
```
Might want to move some functions there.

laanwj commented on src/crypto/sha256_avx2.cpp:
```diff
@@ -0,0 +1,327 @@
+    h = Add(t1, t2);
+}
+
+__m256i inline Read8(const unsigned char* chunk, int offset) {
```
`Read8` and `Write8` appear to read and write values respectively in opposite order, I suppose this is intentional?

unrelated: I also wonder if this could be done with a parallel instruction instead of calling into `ReadBE32`/`WriteBE32` for each component, as we know the host endianness.

sipa replied:
Fixed! Indeed, there exist byte-shuffle intrinsics for SSE4 and AVX2; I've used those instead of individual byteswaps.

kristapsk commented on src/bench/crypto_hash.cpp:
```diff
@@ -52,6 +52,14 @@ static void SHA256_32b(benchmark::State& state)
 
+static void DSHA256_64b(benchmark::State& state)
+{
+    std::vector<uint8_t> in(64 * 1024,0);
```
Shouldn't there be a space between "1024," and "0"?

theuni commented:
Concept ACK!

@sipa Please see the build-system comments on #13203, I didn't realize this was a different PR. I'll follow-up here for non-power changes.

theuni commented on src/crypto/sha256.cpp:
```diff
@@ -465,22 +470,28 @@ bool SelfTest(TransformType tr) {
         TransformD64 = TransformD64Wrapper<sha256_sse4::Transform>;
-        assert(SelfTest(Transform));
-        return "sse4";
+#if defined(ENABLE_SSE41) && !defined(BUILD_BITCOIN_INTERNAL)
```
Just because libbitcoinconsensus doesn't take advantage of 4way? Or some buildsystem limitation?

sipa replied:
Both.

I didn't feel like creating copies of all the architecture-specialized libs for use within libbitcoinconsensus - especially as it doesn't benefit it.

Kick1986 commented:
Great job! 
Any plans to add support for AVX512?

I know there are not so many people with AVX512 CPU's but why not added it from now? 

All Intel CPU's will have support for AVX512 in coming months (mainstream + server).

Kick1986 commented:
Some benchmarks numbers about AVX512 in the article below! 

https://hk.saowen.com/a/218fa211f2b705dc6ebc2691a2b21743ce7cd95585c781d181b3809352bffbe3

laanwj commented:
> I know there are not so many people with AVX512 CPU's but why not add it from now?

That is not how open source development works. A PR is for reviewing the code. Future improvements can be done in future PRs. For example @TheBlueMatt adds support for POWER8 instructions in #13203. I might add ARM intrinsics support at some point.

sipa commented:
@laanwj The algorithm in `MerkleComputation` is actually potentially more efficient (better memory locality) than the one in `ComputeMerkleRoot` now, apart from the fact that it doesn't take advantage of multi-way hashing. I'd like to keep it around for a bit and see if I can adapt it to use multi-way instead, in which case it could be used for everything. I can also move it and move it back if used. What do you think?

@Kick1986 AVX512 is cool, but it's low-impact (even machines that support it right now do it with reduced clock rate), and I'm unable to benchmark it. Feel free to add in follow-up work yoursel, though.

Kick1986 commented:
@sipa
Thanks for answering me!
You are right about AVX512 clock speed, it will go mainstream sometime 2020.
I will do my best to add AVX512 to BTC code before that.
Thanks

sipa committed:
[MOVEONLY] Move unused Merkle branch code to tests

sipa commented:
@laanwj I moved the unused Merkle branch functions to the test code.

@theuni I've addressed some of your build system comments from #13203, but left the `ENABLE_AVX2` and `ENABLE_SSE41` macros, because there isn't a clean platform independent way to use compiler defines. `__AVX2__` exists in both GCC and MSVC, but there is no `__SSE4_1__` in MSVC (and worse, there is no way to test for SSE4 at all there; you have to test whether the FP code is x87 based or SSE based).

@kristapsk Ok, I've added a space.

sipa commented:
@gmaxwell Yes, perhaps. But the impact of that would be pretty low at best, as every system in which that optimized 1-way code can be used also supports 4-way 64-byte optimized code already; meaning such an implementation would only be used for up to 3 of the last hashes in each level of a Merkle tree.

sipa commented:
@theuni Could you have a look at the build system changes again?

theuni reviewed (approved):
utACK, nothing major to complain about. Admittedly I'm relying on tests to check the new sha2 code itself, I only glanced. Very nice work :)

Might it be helpful to add a non-double 4way/8way function as well? That would allow (for example) batched txid calculations where the first iteration is lazy and variable-sized as it is now, but 4way/8way could be used for the fixed-size second iterations.

sipa commented on src/crypto/sha256.cpp:
```diff
@@ -61,73 +61,73 @@ void Transform(uint32_t* s, const unsigned char* chunk, size_t blocks)
-        Round(d, e, f, g, h, a, b, c, 0xa4506ceb, w13 += sigma1(w11) + w6 + sigma0(w14));
-        Round(c, d, e, f, g, h, a, b, 0xbef9a3f7, w14 + sigma1(w12) + w7 + sigma0(w15));
-        Round(b, c, d, e, f, g, h, a, 0xc67178f2, w15 + sigma1(w13) + w8 + sigma0(w0));
+        Round(a, b, c, d, e, f, g, h, 0x428a2f98 + (w0 = ReadBE32(chunk + 0)));
```
Fixed (also in SSE4 and AVX2 code).

sipa commented on src/consensus/merkle.cpp:
```diff
@@ -131,9 +131,23 @@ static void MerkleComputation(const std::vector<uint256>& leaves, uint256* proot
 }
 
 uint256 ComputeMerkleRoot(const std::vector<uint256>& leaves, bool* mutated) {
```
Done.

sipa commented:
Addressed @theuni's nits.

Also:

> Might it be helpful to add a non-double 4way/8way function as well? That would allow (for example) batched txid calculations where the first iteration is lazy and variable-sized as it is now, but 4way/8way could be used for the fixed-size second iterations.

Yes, but it's significantly more complicated. You need to schedule multiple variable length things in groups of 64 bytes, interspersed with padding. Not going to do that in this PR. Also, @jl2012 suggested writing an optimized 32-byte-input single-SHA256 for use in the second half of double-SHA256 computations, which could help there as well.

DesWurstes commented:
In the future, please don’t spend time coding AVX/AVX2 SHA256. SHA-NI specialized instruction set has SHA2 opcodes, which is the possible most efficient way. Besides, it was released before SSE4.2 and AVX, so there are compatible more processors.

If anyone is interested: https://github.com/01org/isa-l_crypto/blob/master/sha256_mb/sha256_ni_x2.asm

sipa commented:
@DesWurstes I'm perfectly capable of deciding for myself what I find interesting, thanks. Hardly any systems today support SHA-NI (only very recent low-power Intel CPUs, and AMD Ryzen), while AVX2 is available on all Intel chips since 2013 and AMD chips since 2015.

laanwj commented:
@DesWurstes Apparently you didn't bother to read the posts at all before replying, as this came up before. People decide for themselves what to work on. If you think something else is more important, you can submit your own pull request.

sipa commented:
@DesWurstes https://github.com/bitcoin/bitcoin/pull/13386

laanwj commented:
utACK 4defdfab94504018f822dc34a313ad26cedc8255
Verified that FreeBSD+OpenBSD builds still pass.

droark commented:
After-the-fact tACK

The code's running nicely on my AVX2-enabled machine. Good job!

I do have one question. Is this going to affect Gitian builds for people who are on older machines? Maybe I'm missing something silly but it seems like this could cause mismatches. I remember that the SSE4-enabled SHA-256 code initially required --enable-experimental-asm and was eventually enabled by default.

Thanks.

sipa commented:
@droark It does require an AVX2 compatible compiler, but GCC 4.7 is sufficient for that (which we already used for release builds). For 0.17 we're likely going to upgrade the build environment to Ubuntu 18.04 LTS as well.

droark commented:
@sipa - Thanks. I had a bad understanding of compilers from long ago that I somehow never shook.

rollmeister commented:
Can you please provide any documentation as to how you calculated the custom constants in Transform 2? I would like to adapt it for a 112 byte input.
